File: Java/ValidParentheses.java
import java.util.Stack;

public class ValidParentheses {

    public static boolean isValid(String s) {
        // Stack to keep track of opening brackets
        Stack<Character> stack = new Stack<>();

        // Iterate through each character in the string
        for (char c : s.toCharArray()) {
            // If it's an opening bracket, push to stack
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            }
            // If it's a closing bracket, check if it matches the top of the stack
            else if (c == ')' && !stack.isEmpty() && stack.peek() == '(') {
                stack.pop();
            } else if (c == '}' && !stack.isEmpty() && stack.peek() == '{') {
                stack.pop();
            } else if (c == ']' && !stack.isEmpty() && stack.peek() == '[') {
                stack.pop();
            }
            // If it doesn't match or the stack is empty, it's invalid
            else {
                return false;
            }
        }

        // If the stack is empty, all brackets were matched, so the string is valid
        return stack.isEmpty();
    }

    public static void main(String[] args) {
        String test = "([{}])";
        if (isValid(test)) {
            System.out.println("The string is valid.");
        } else {
            System.out.println("The string is invalid.");
        }
    }
}

File: C++/ValidParentheses.cpp
#include <iostream>
#include <stack>
#include <string>

class Solution {
public:
    bool isValid(const std::string& s) {
        std::stack<char> stack;
        
        for (char c : s) {
            switch (c) {
                case '(': case '[': case '{':
                    stack.push(c);
                    break;
                case ')':
                    if (stack.empty() || stack.top() != '(') return false;
                    stack.pop();
                    break;
                case ']':
                    if (stack.empty() || stack.top() != '[') return false;
                    stack.pop();
                    break;
                case '}':
                    if (stack.empty() || stack.top() != '{') return false;
                    stack.pop();
                    break;
                default:
                    return false; // Invalid character
            }
        }
        
        return stack.empty();
    }
};

// Example usage
int main() {
    Solution solution;
    std::string input;

    // Test with a few examples
    std::cout << "Enter a string of parentheses: ";
    std::cin >> input;

    if (solution.isValid(input)) {
        std::cout << "The string is valid.\n";
    } else {
        std::cout << "The string is not valid.\n";
    }

    return 0;
}
